File: teamcity-vsix-gallery-server/src/main/teamcity/vsix/index/VsixPackageStructureVisitor.java
package teamcity.vsix.index;

import com.intellij.openapi.diagnostic.Logger;
import jetbrains.buildServer.serverSide.artifacts.BuildArtifact;
import jetbrains.buildServer.util.FileUtil;
import org.jdom.Element;
import org.jdom.JDOMException;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import teamcity.vsix.feed.FeedConstants;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Collection;
import java.util.zip.ZipEntry;
import java.util.zip.ZipInputStream;

public class VsixPackageStructureVisitor {

    private static final Logger LOG = Logger.getInstance("teamcity.vsix");

    @NotNull
    private final Collection<VsixPackageStructureAnalyser> myAnalysers;

    public VsixPackageStructureVisitor(@NotNull Collection<VsixPackageStructureAnalyser> analysers) {
        myAnalysers = analysers;
    }

    public void visit(@NotNull BuildArtifact artifact) throws PackageLoadException {
        if(myAnalysers.isEmpty()) return;
        ZipInputStream zipInputStream = null;
        InputStream stream = null;
        final String vsixPackageName = artifact.getName();
        try {
            stream = artifact.getInputStream();
            zipInputStream = new ZipInputStream(new BufferedInputStream(stream));
            ZipEntry zipEntry;
            while ((zipEntry = zipInputStream.getNextEntry()) != null) {
                if(zipEntry.isDirectory()) continue;
                final String zipEntryName = zipEntry.getName();
                if (zipEntryName.endsWith(FeedConstants.VSIXMANIFEST_FILE_EXTENSION)) {
                    LOG.info(String.format("Manifest file found on path %s in VSIX package %s", zipEntryName, vsixPackageName));
                    final VsixManifestContent vsixManifestContent = readVsixManifestContent(zipInputStream);
                    if (vsixManifestContent == null)
                        LOG.warn("Failed to read .vsixmanifest file content from NuGet package " + vsixPackageName);
                    else {
                        for(VsixPackageStructureAnalyser analyser : myAnalysers){
                            analyser.analyzeVsixManifest(vsixManifestContent);
                        }
                    }
                    zipInputStream.closeEntry();
                }
            }
        } catch (IOException e) {
            LOG.warn("Failed to read content of NuGet package " + vsixPackageName);
            if(zipInputStream != null){
                try {
                    zipInputStream.close();
                } catch (IOException ex) {
                    //NOP
                }
            }
        } finally {
            FileUtil.close(stream);
        }
    }

    @Nullable
    private VsixManifestContent readVsixManifestContent(final ZipInputStream finalZipInputStream) throws IOException {
        try {
            final Element document = FileUtil.parseDocument(new InputStream() {
                @Override
                public int read() throws IOException {
                    return finalZipInputStream.read();
                }

                @Override
                public void close() throws IOException {
                    //do nothing, should avoid stream closing by xml parse util
                }
            }, false);
            return new VsixManifestContent(document);
        } catch (JDOMException e) {
            LOG.debug(e);
            return null;
        }
    }
}

File: teamcity-vsix-gallery-server/src/main/teamcity/vsix/index/VsixPackageStructureVisitor.kt
package teamcity.vsix.index


import com.intellij.openapi.diagnostic.Logger
import jetbrains.buildServer.serverSide.artifacts.BuildArtifact
import jetbrains.buildServer.util.FileUtil
import org.jdom.JDOMException
import teamcity.vsix.feed.VSIXMANIFEST_FILE_EXTENSION
import java.io.BufferedInputStream
import java.io.IOException
import java.io.InputStream
import java.util.zip.ZipInputStream

public class VsixPackageStructureVisitor(private val analysers: Collection<VsixPackageStructureAnalyser>) {

    val log = Logger.getInstance("teamcity.vsix")

    public fun visit(artifact: BuildArtifact) {
        if (analysers.isEmpty()) {
            return
        }
        // TODO: Refactor all this to a more functional style
        val vsixPackageName = artifact.getName()
        val stream = artifact.getInputStream()
        // TODO: this is not nice!
        var zipInputStream: ZipInputStream? = null
        try {
            zipInputStream = ZipInputStream(BufferedInputStream(stream))
            var zipEntry = zipInputStream?.getNextEntry()
            while (zipEntry != null) {
                if (zipEntry?.isDirectory()!!) {
                    continue
                }
                val zipEntryName = zipEntry?.getName()!!
                if (zipEntryName.endsWith(VSIXMANIFEST_FILE_EXTENSION)) {
                    log.info("Manifest file found on path $zipEntryName in VSIX package $vsixPackageName")
                    val vsixManifestContent = readVsixManifestContent(zipInputStream!!)
                    if (vsixManifestContent == null) {
                        log.warn("Failed to read .vsixmanifest file content from VSIX package $vsixPackageName")
                    } else {
                        analysers.forEach {
                            it.analyzeVsixManifest(vsixManifestContent)
                        }
                    }
                    zipInputStream?.closeEntry()
                }
            }
        } catch (ioException: IOException) {
            log.warn("Failed to read content of VSIX package " + vsixPackageName);
            log.warn(ioException.toString());
            try {
                zipInputStream?.close()

            } catch (exception: IOException) {

            }
        } finally {
            FileUtil.close(stream)
        }
    }

    private fun readVsixManifestContent(zipInputStream: ZipInputStream): VsixManifestContent? {
        try {
            val document = FileUtil.parseDocument(object: InputStream(){
                override fun close() {
                    // Do nothing
                }

                override fun read(): Int {
                    return zipInputStream.read()
                }
            }, false)
            return VsixManifestContent(document)
        } catch (exception: JDOMException) {
            log.debug(exception)
            return null
        }
    }
}